keps/sig-node/753-keystone-containers/README.md

```diff
@@ -0,0 +1,815 @@
+
+Another motivation for this KEP is to take small steps towards the sidecar containers that will help a lot of users without including complicated logic and changing critical kubelet code paths.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Pod termination by kubelet if application containers have successfully run to completion and exited.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+- Handling Container startup/shutdown ordering for application and sidecar containers
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+A pod annotation can be specified by users which contains a list of all core application containers in the pod. Depending on the user feedback and how we want to proceed with the sidecar proposal this annotation can be modified or promoted to an API field. 
+
+The annotation is used by kubelet to make a decision on pod termination based on the status of application containers and pod restart policy.
+
+- if all application containers exited successfully and the restart policy of the pod is "never" or "onFailure" then terminate the pod.
+
+- if all app containers are failed and the restart policy is "never" then terminate the pod. 
+If failure logs are concerned in this case (considering the logging container is sidecar) then the sidecar container will only be terminated when all app containers are in the failed state, so it must have captured the failure logs.
+
+The following are existing behaviors just mentioned here for clarity:
+
+- if all application containers exited successfully and the restart policy of the pod is "always" then restart the containers.
+
+- if all app containers are failed and the restart policy is "onfailure", then kubelet will keep trying to restart containers, and no point in terminating the pod
+
+Since container annotations are not supported to be added by an end user, We need to use pod annotation to specify application containers.
+
+There are few options for annotations:
+
+1. Specifying a list of keystone or sidecar containers in one annotation
+   e.g. `alpha.node.kubernetes.io/sidecars:{list of sidecar containers}` or
+        `alpha.node.kubernetes.io/keystone:{list of keystone containers}`
+2. One annotation per container
+   e.g. `alpha.sidecars.kubernetes.io/<container name>:string`
```

Comment:
I'm really NOT a fan of this approach.  It sets a bad precedent.

Reply:
I don't have lots of context, Annotation approach has been discussed in very old threads https://kubernetes.slack.com/archives/C0BP8PW9G/p1620748257450100?thread_ts=1620304727.344300&cid=C0BP8PW9G
cc @dims @derekwaynecarr

Reply:
@thockin do you mean because you are not sure more than one keystone container is needed or because it is an annotation instead of a field?

Reply:
All of the above.

Annotations were designed for relatively low cardinality.  In this model, each pod will be small-N, but in aggregate the number of unique annotations in a cluster will be unbounded.  I'm not sure that anyone currently tries to collect all such annotations in a meaningful way (yet) but I have, for example, wanted to know "who me all the annotations used in all of our clusters".  Putting variables in the key means I have to parse keys for semantic equivalence.

I'm not convinced we need more than one keystone (though it seems possible, I admit).

But maybe most importantly - being an annotation doesn't make it less "real" or less subject to our deprecation rules.  It just makes it harder to use.

Reply:
Cool, let's go for an pod.spec field, then! :)

Reply:
Oh, that was before I read your comment on the PR itself. Then, by [this comment](https://github.com/kubernetes/enhancements/pull/2869#issuecomment-999020318) I understand you are ok with an annotation, and that it seems better.

So, keeping it as annotation for now :)